statefun-flink/statefun-flink-core/src/main/java/org/apache/flink/statefun/flink/core/reqreply/RequestReplyFunction.java

```diff
@@ -125,8 +126,44 @@ private void onAsyncResult(
       sendToFunction(context, batch);
       return;
     }
-    InvocationResponse invocationResult = unpackInvocationOrThrow(context.self(), asyncResult);
-    handleInvocationResponse(context, invocationResult);
+    if (asyncResult.failure()) {
+      throw new IllegalStateException(
+          "Failure forwarding a message to a remote function " + context.self(),
+          asyncResult.throwable());
+    }
+
+    final Either<InvocationResponse, IncompleteInvocationContext> response =
+        unpackResponse(asyncResult.value());
+    if (response.isRight()) {
+      handleIncompleteInvocationContextResponse(context, response.right(), asyncResult.metadata());
+    } else {
+      handleInvocationResultResponse(context, response.left());
+    }
+  }
+
+  private static Either<InvocationResponse, IncompleteInvocationContext> unpackResponse(
```

Comment:
I've tested this before and it looks like if a `oneof` field was not set, then getting either one of the type values would return the type's default instance, i.e.:

if `fromFunction.hasInvocationResult()` is `false`, then the object returned by `fromFunction.getInvocationResult` equals `InvocationResult.getDefaultInstance()`. The get method would not throw.

Reply:
But I guess it won't hurt to make this explicit in the code as you suggested, so that the behaviour is clear:

```
if (fromFunction.hasIncompleteInvocationContext()) {
    return Either.Right(fromFunction.getIncompleteInvocationContext());
}

if (fromFunction.hasInvocationResult()) {
    return Either.Left(fromFunction.getInvocationResult());
}

// function had no side effects
return Either.Left(InvocationResult.getDefaultInstance());
```

does this look good to you?

Reply:
Very good, yes!